src/dfx/src/lib/operations/cycles_ledger.rs

```diff
@@ -170,3 +184,89 @@ pub async fn send(
 
     Ok(block_index)
 }
+
+#[context("Failed to create canister via cycles ledger.")]
+pub async fn create_with_cycles_ledger(
+    agent: &Agent,
+    with_cycles: Option<u128>,
+    from_subaccount: Option<Subaccount>,
+    settings: DfxCanisterSettings,
+    cycles_ledger_canister_id: Principal,
+) -> DfxResult<Principal> {
+    #[derive(CandidType, Clone, Debug)]
+    // TODO(FI-1022): Import types from cycles ledger crate once available
+    struct CreateCanisterArgs {
+        pub from_subaccount: Option<icrc_ledger_types::icrc1::account::Subaccount>,
+        pub created_at_time: Option<u64>,
+        pub amount: u128,
+        pub creation_args: Option<CmcCreateCanisterArgs>,
+    }
+    #[derive(CandidType, Clone, Debug)]
+    struct CmcCreateCanisterArgs {
+        pub subnet_selection: Option<SubnetSelection>,
+        pub settings: Option<CanisterSettings>,
+    }
+    #[derive(CandidType, Clone, Debug)]
+    #[allow(dead_code)]
+    enum SubnetSelection {
+        /// Choose a random subnet that satisfies the specified properties
+        Filter(SubnetFilter),
+        /// Choose a specific subnet
+        Subnet { subnet: Principal },
+    }
+    #[derive(CandidType, Clone, Debug)]
+    struct SubnetFilter {
+        pub subnet_type: Option<String>,
+    }
+    #[derive(CandidType, Clone, Debug, Deserialize, Error)]
+    enum CreateCanisterError {
+        #[error("Insufficient funds. Current balance: {balance}")]
+        InsufficientFunds { balance: u128 },
+        #[error("Local clock too far behind.")]
+        TooOld,
+        #[error("Local clock too far ahead.")]
+        CreatedInFuture { ledger_time: u64 },
+        #[error("Cycles ledger temporarily unavailable.")]
+        TemporarilyUnavailable,
+        #[error("Duplicate of block {duplicate_of}.")]
+        Duplicate { duplicate_of: Nat },
+        #[error("Cycles ledger failed to create canister: {error}")]
+        FailedToCreate {
+            fee_block: Option<Nat>,
+            refund_block: Option<Nat>,
+            error: String,
+        },
+        #[error("Ledger error {error_code}: {message}")]
+        GenericError { error_code: Nat, message: String },
+    }
+    #[derive(Deserialize, CandidType, Clone, Debug, PartialEq, Eq)]
+    struct CreateCanisterSuccess {
+        pub block_id: Nat,
+        pub canister_id: Principal,
+    }
+
+    let cycles = with_cycles.unwrap_or(CANISTER_CREATE_FEE + CANISTER_INITIAL_CYCLE_BALANCE);
+    let result = agent
```

Comment:
Please set `created_at_time`, so that it's an idempotent operation, and wrap this call in retry logic.  This will reduce the likelihood of burning cycles creating a canister but not storing the canister id.

Reply:
I'm setting and logging the `created_at_time` now so the result will be recoverable. I don't think it makes sense to add a loop because there is no error for which a retry really makes sense (except for whatever the agent can retry on its own).

Reply:
The case where retry makes sense is the case where the replica executed the operation, but the client didn't find out about it.

Reply:
Though it is possible to handle all of this by making the user retry their command (this time specifying `--created-at-time` and altering the command to create or deploy a single canister), in a case where we have an idempotent operation, in my opinion it's better to handle that retry in code.

As for determine when to retry, you are correct: it is difficult.  A guess is `AgentError::TimeoutWaitingForResponse() | AgentError::TransportError(_)`, here: https://github.com/dfinity/sdk/blob/master/src/dfx/src/lib/retryable.rs#L6.

Regarding "whatever the agent can retry on its own", the agent does not retry `TransportError` because it can't. 
 (See https://github.com/dfinity/agent-rs/blob/main/ic-agent/src/agent/http_transport/reqwest_transport.rs#L88 for an example, and up the call stack).  It can't because after sending the initial request, there exists a class of errors for which it can't know if the server received the request and is executing it or not, and the agent doesn't know if the request is idempotent or not.

Relying on the agent's retry mechanisms also means relying on those mechanisms to be correct.  They may or may not be.  For example, Agent's [fn wait](https://github.com/dfinity/agent-rs/blob/main/ic-agent/src/agent/mod.rs#L673) doesn't retry if [fn poll](https://github.com/dfinity/agent-rs/blob/main/ic-agent/src/agent/mod.rs#L632) returns an Err, and fn poll will return an error right away in the case of connection timeout when making the read_state call.

Given the above, what do you think?

Reply:
The idempotency argument makes sense. I'll use `retryable` for now and if we find a better heuristic we can always impove it